File: cpp/144BinaryTreePreorderTraversal.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

/**
 * Definition for a binary tree node.
 */
 struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode() : val(0), left(nullptr), right(nullptr) {}
     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> output;

        /*
         * recursive solution
        function<void(TreeNode*)> traverse = [&output, &traverse](TreeNode* node) {
            if (node == nullptr) {
                return;
            }

            output.push_back(node->val);
            traverse(node->left);
            traverse(node->right);
        };

        traverse(root);
        */

        //iterative solution
        stack<TreeNode*> deferedNodes;
        while (root != nullptr) {
            output.push_back(root->val);
            if (root->left != nullptr) {
                if (root->right != nullptr) {
                    deferedNodes.push(root->right);
                }
                root = root->left;
            } else {
                if (root->right != nullptr) {
                    root = root->right;
                } else {
                    if (deferedNodes.empty()) {
                        break;
                    } else {
                        root = deferedNodes.top();
                        deferedNodes.pop();
                    }
                }
            }
        }

        return output;
    }
};


int main(){
    Solution s;
    TreeNode t1(3);
    TreeNode t2(2, &t1, nullptr);
    TreeNode root(1, nullptr, &t2);
    auto output = s.preorderTraversal(&root);

    auto output1 = s.preorderTraversal(nullptr);

    TreeNode root2(1);
    auto output2 = s.preorderTraversal(&root2);

    TreeNode t3(2);
    TreeNode root3(1, &t3, nullptr);
    auto output3 = s.preorderTraversal(&root3);


    TreeNode t4(2);
    TreeNode root4(1, nullptr, &t4);
    auto output4 = s.preorderTraversal(&root4);

    TreeNode t5(1);
    TreeNode t6(2);
    TreeNode root5(3, &t5, &t6);
    auto output5 = s.preorderTraversal(&root5);

    cout << "*************output*************" << endl;
    printContainer(output);
    printContainer(output1);
    printContainer(output2);
    printContainer(output3);
    printContainer(output4);
    printContainer(output5);

    return 0;
}

File: go/144BinaryTreePreorderTraversal.go
package main

import (
	"container/list"
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preorderTraversal(root *TreeNode) []int {
	output := []int{}

	/*
		//recursive solution
		var traverse func(*TreeNode, *[]int)
		traverse = func(node *TreeNode, nodeVal *[]int) {
			if node == nil {
				return
			}

			*nodeVal = append(*nodeVal, node.Val)

			traverse(node.Left, nodeVal)
			traverse(node.Right, nodeVal)
		}

		traverse(root, &output)
	*/

	//iterative solution
	staPendingRightNodes := list.New()
	curNode := root

	for curNode != nil {
		output = append(output, curNode.Val)

		if curNode.Left != nil {
			if curNode.Right != nil {
				staPendingRightNodes.PushFront(curNode.Right)
			}
			curNode = curNode.Left
		} else {
			if curNode.Right != nil {
				curNode = curNode.Right
			} else {
				if staPendingRightNodes.Front() != nil {
					curNode = staPendingRightNodes.Front().Value.(*TreeNode)
					staPendingRightNodes.Remove(staPendingRightNodes.Front())
				} else { // the end of traversal
					curNode = nil
				}
			}
		}
	}

	return output
}

func main() {
	output := preorderTraversal(
		&TreeNode{
			Val:  1,
			Left: nil,
			Right: &TreeNode{
				Val: 2,
				Left: &TreeNode{
					Val:   3,
					Left:  nil,
					Right: nil,
				},
				Right: nil,
			},
		})
	output1 := preorderTraversal(nil)
	output2 := preorderTraversal(
		&TreeNode{
			Val:   1,
			Left:  nil,
			Right: nil,
		})
	output3 := preorderTraversal(
		&TreeNode{
			Val: 1,
			Left: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
			Right: nil,
		})
	output4 := preorderTraversal(
		&TreeNode{
			Val:  1,
			Left: nil,
			Right: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
		})
	output5 := preorderTraversal(
		&TreeNode{
			Val: 3,
			Left: &TreeNode{
				Val:   1,
				Left:  nil,
				Right: nil,
			},
			Right: &TreeNode{
				Val:   2,
				Left:  nil,
				Right: nil,
			},
		})

	fmt.Println("\n******** output *********\n", output, output1, output2, output3, output4, output5)
}
